readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/writer/ProtoParquetWriterWithOffset.java

```diff
@@ -74,26 +79,76 @@ public Path close() throws IOException {
+
+        final Path finalPath = new Path(finalHdfsDir, fileNamer.computePath(dayStartTime, lastIndex, latestOffset));
 
         FileSystemUtils.ensureDirectoriesExist(Collections.singleton(finalPath.getParent()), fs);
 
-        if (!fs.rename(temporaryHdfsPath, finalPath)) {
-            throw new IOException(String.format("Failed to commit %s (from %s)",
-                    finalPath.toUri(), temporaryHdfsPath));
+        if (lastAvailableFinalPath.isPresent()) {
+            long blockSize = fs.getStatus(lastAvailableFinalPath.get()).getUsed();
+            if (blockSize > fsBlockSize) {
+                moveToFinalPath(temporaryHdfsPath, finalPath);
+            } else {
+                mergeToFinalPath(lastAvailableFinalPath.get(), finalPath);
+            }
+        } else {
+            moveToFinalPath(temporaryHdfsPath, finalPath);
         }
 
         LOGGER.info("Committed {} (from {})", finalPath.toUri(), temporaryHdfsPath);
 
         return finalPath;
     }
 
+    protected void moveToFinalPath(Path tempPath, Path finalPath) throws IOException {
+        if (!fs.rename(tempPath, finalPath)) {
+            throw new IOException(String.format("Failed to commit %s (from %s)",
+                    finalPath.toUri(), tempPath));
+        }
+    }
+
+    protected void mergeToFinalPath(Path lastAvailableFinalPath, Path finalPath) throws IOException {
+        MessageType schema = ParquetFileReader.open(fs.getConf(), lastAvailableFinalPath).getFileMetaData().getSchema();
+        if (!checkSchemaEquality(schema)) {
+            LOGGER.warn("Schema between last available final file ({}) and temp file ({}) are not identical. We can't merge them",
+                    lastAvailableFinalPath, temporaryHdfsPath);
+            moveToFinalPath(temporaryHdfsPath, finalPath);
+        } else {
+            Path mergedTempFile = new Path(temporaryHdfsPath.toString() + ".merged");
+
+            if (fs.isFile(mergedTempFile)) fs.delete(mergedTempFile, false);
+
+            ParquetFileWriter writerPF = new ParquetFileWriter(fs.getConf(), schema, mergedTempFile);
+            writerPF.start();
+            writerPF.appendFile(fs.getConf(), lastAvailableFinalPath);
+            writerPF.appendFile(fs.getConf(), temporaryHdfsPath);
+            writerPF.end(EMPTY_METADATA);
+
+            moveToFinalPath(mergedTempFile, lastAvailableFinalPath);
+            try {
+                fs.delete(temporaryHdfsPath, false);
+                // This file is in a temp folder that should be deleted at exit so we should not throw exception here
+            } catch (IOException ignored) {
+            }
+        }
+    }
+
+    private Boolean checkSchemaEquality(MessageType schema) throws IOException {
+        MessageType schema2 = ParquetFileReader.open(fs.getConf(), temporaryHdfsPath).getFileMetaData().getSchema();
+
+        return schema.getColumns().size() == schema2.getColumns().size() &&
```

Comment:
why not schema.equals(schema2) ?
or schema.checkContains(schema2) || schema2.checkContains(schema) ?

Reply:
checkContains -> return void so for sure I can't use it
Will check for schema.equals(schema2)